hypothesis/client
# Check annotator-sidebar version consistency and sidebar origin on startup

This PR adds checks for three potential problems with the environment when the sidebar app starts. If the checks fail client startup is allowed to continue, but we log a warning in the browser console and disable Sentry error reporting to reduce noise.

The problems that are checked for currently are:

- Sidebar and annotator versions do not match, which _may_ cause problems with communication between the sidebar and other frames
- The sidebar is loaded in a sandboxed iframe, indicated by `window.origin` being "null"
- The sidebar is served from the wrong origin

The code changes are:

- Add an `origin` parameter to the `#config` fragment in the sidebar/notebook app launch URLs which indicates the expected origin of the sidebar app.
- Add a `checkEnvironment` function in `src/sidebar/config/check-env.js` which performs the checks and returns true they passed or logs warnings and returns false otherwise.
- In `src/sidebar/index.js`, call `checkEnvironment` and disable Sentry reporting if it fails.

## Background

When deploying the new inter-frame communication code we saw errors [1] setting up communication between the sidebar and host frames caused by:

 1. Mismatches between the version of the client used in the annotator and the sidebar
 2. The sidebar being served from an unexpected origin (eg. due to use of a web proxy) or an opaque origin (eg. due to the host being loaded in a sandboxed iframe). This is generally outside of our control.

We should be able to reduce occurrences of (1) by changing caching settings on the client's boot script. See https://github.com/hypothesis/client/pull/4050. However there are also situations out of our control where it can happen, eg. if a web page embedding Hypothesis is saved as an archive using certain tools, or a proxy overrides the boot script's caching headers.

Situation (2) is outside of our control and we are not going to relax origin checks in the login process as they are essential for security.

[1] See https://github.com/hypothesis/client/issues/3986#issuecomment-989685057 for more details.

## Testing

1. In a normal startup (ie. not in one of the environments where the client is expected to fail), no new warnings should be logged.
2. Change the `checkEnvironment` call near the top of `sidebar/index.js` to either of the following variants. The client should log a warning on startup:

```js
// Simulate sidebar being loaded from another origin
const envOk = checkEnvironment({ origin: 'https://hypothesis.someproxy.edu', location });

// Simulate sidebar being loaded in sandboxed iframe
const envOk = checkEnvironment({ origin: 'null', location });
```
3. Change the `appConfig.version` initialization in `src/annotator/config/app.js` to `appConfig.version = '1.0.0-other-version'` and reload the page. You should see a warning about the version mismatch.
